python/ray/autoscaler/v2/instance_manager/instance_manager.py

```diff
@@ -0,0 +1,296 @@
+        if len(ids_to_updates) == 0:
+            to_update_instances, version = {}, request.expected_version
+        else:
+            to_update_instances, version = self._instance_storage.get_instances(
+                ids_to_updates.keys()
+            )
+            if request.expected_version >= 0 and request.expected_version != version:
+                reply = UpdateInstanceManagerStateReply()
+                reply.status = Status.OK
+                reply.version = version
+                return reply
+
+        for instance in to_update_instances.values():
+            update = ids_to_updates[instance.instance_id]
+            InstanceUtil.set_status(instance, new_ray_status=update.new_ray_status)
+
+        # 2. Handle new instances to start
+        new_instances = []
+        for launch_request in request.launch_requests:
+            instances = self._create_instances(launch_request)
+            new_instances.extend(instances)
+
+        expected_version = (
+            request.expected_version if request.expected_version else None
+        )
+
+        # 3. Update the instance storage.
+        ok, version = self._instance_storage.batch_upsert_instances(
+            new_instances + list(to_update_instances.values()),
+            expected_version,
+        )
+
+        if ok:
+            reply.status = Status.OK
+        else:
+            if version != expected_version:
+                reply.status = Status.VERSION_MISMATCH
+            else:
+                reply.status = Status.UNKNOWN_ERROR
+
+        reply = UpdateInstanceManagerStateReply()
+        reply.state.CopyFrom(self.get_instance_manager_state().state)
+        reply.version = reply.state.version
+        return reply
+
+    def get_instance_manager_state(self) -> GetInstanceManagerStateReply:
+        """
+        Returns:
+            GetInstanceManagerStateReply: The current instance manager state.
+
+        Override:
+            Overrides the base class method.
+        """
+        reply = GetInstanceManagerStateReply()
+        reply.state.CopyFrom(self._get_instance_manager_state())
+        reply.status = Status.OK
+        return reply
+
+    def _get_instance_manager_state(self) -> InstanceManagerState:
+        """
+        Utility function to get the current instance manager state.
+        """
+        instances, version = self._instance_storage.get_instances()
+        state = InstanceManagerState()
+        state.version = version
+        state.instances.extend(instances.values())
+        return state
+
+    def _create_instances(self, request: LaunchRequest) -> List[Instance]:
+        instances = []
+        instance_type = request.instance_type
+        assert instance_type in self._instance_configs, (
+            f"instance type {instance_type} not found in "
+            f"instance configs {self._instance_configs}"
+        )
+        for _ in range(request.count):
+            instance = self.new_instance(
```

Comment:
is this instance actually creating real instances? Otherwise, we should probably mention in the comment it is just in-memory data structure